trieste/space.py

```diff
@@ -128,20 +134,22 @@ def __mul__(self, other: DiscreteSearchSpace) -> DiscreteSearchSpace:
                    [2, 3, 4, 5, 6],
                    [2, 3, 7, 8, 9]], dtype=int32)
 
-        :param other: :class:`DiscreteSearchSpace`.
-        :return: the new combined :class:`DiscreteSearchSpace`
-        :raise TypeError: If the lhs and rhs :class:`DiscreteSearchSpace` points have different types.
+        :param other: A :class:`DiscreteSearchSpace` with :attr:`points` of the same dtype as this
+            search space.
+        :return: The Cartesian product of the two :class:`DiscreteSearchSpace`\ s.
+        :raise TypeError: If one :class:`DiscreteSearchSpace` has :attr:`points` of a different
+            dtype to the other.
         """
-        if self._points.dtype is not other._points.dtype:
+        if self._points.dtype is not other.points.dtype:
             return NotImplemented
 
-        N = self._points.shape[0]
-        M = other._points.shape[0]
-        tile_self = tf.tile(tf.expand_dims(self.points, 1), [1, M, 1])
-        tile_dss = tf.tile(tf.expand_dims(other.points, 0), [N, 1, 1])
-        cartesian_product = tf.concat([tile_self, tile_dss], axis=2)
+        dim_self = self._points.shape[-1]
+        dim_other = other.points.shape[-1]
```

Comment:
I'd prefer keeping N, let's say N0 and N1, from a context it is clear what you mean by them.

Reply:
hmmm ... i changed it from that because i didn't find it clear. I'll look for a way that has the best of both

Reply:
i've removed them and just named their product